File: bnf/python/ds_bnf/unparse.py
"""
Unparse: Convert from lisp-like Ds syntax to traditional Dsp syntax
"""

from antlr4 import InputStream, CommonTokenStream
from .generated.DsLexer import DsLexer
from .generated.DsParser import DsParser
from .generated.DsVisitor import DsVisitor


class UnparseVisitor(DsVisitor):
    """Visitor to convert from lisp-like Ds syntax to traditional Dsp syntax"""

    def visitRule_pool(self, ctx):
        rules = ctx.rule_()
        if not rules:
            return ""
        return "\n".join(self.visit(r) for r in rules)

    def visitRule(self, ctx):
        terms = ctx.term()
        if not terms:
            return ""

        result = [self.visit(t) for t in terms]
        conclusion = result.pop()

        if not result:
            return conclusion

        return ", ".join(result) + " -> " + conclusion

    def visitSymbol(self, ctx):
        return ctx.SYMBOL().getText()

    def visitSubscript(self, ctx):
        terms = ctx.term()
        if not terms:
            return ""

        base = self.visit(terms[0])
        indices = ", ".join(self.visit(t) for t in terms[1:])
        return f"{base}[{indices}]"

    def visitFunction(self, ctx):
        terms = ctx.term()
        if not terms:
            return ""

        func = self.visit(terms[0])
        args = ", ".join(self.visit(t) for t in terms[1:])
        return f"{func}({args})"

    def visitUnary(self, ctx):
        op = ctx.SYMBOL().getText()
        operand = self.visit(ctx.term(0))
        return f"{op} {operand}"

    def visitBinary(self, ctx):
        op = ctx.SYMBOL().getText()
        left = self.visit(ctx.term(0))
        right = self.visit(ctx.term(1))
        return f"({left} {op} {right})"


def unparse(input_text: str) -> str:
    """
    Convert from lisp-like Ds syntax to traditional Dsp syntax

    Args:
        input_text: Input text in Ds syntax

    Returns:
        Output text in Dsp syntax
    """
    input_stream = InputStream(input_text)
    lexer = DsLexer(input_stream)
    token_stream = CommonTokenStream(lexer)
    parser = DsParser(token_stream)
    tree = parser.rule_pool()
    visitor = UnparseVisitor()
    return visitor.visit(tree)

File: bnf/javascript/src/unparse.ts
import antlr4 from 'antlr4';
import DsLexer from './generated/DsLexer.js';
import DsParser from './generated/DsParser.js';
import DsVisitor from './generated/DsVisitor.js';

/**
 * Visitor to convert from lisp-like Ds syntax to traditional Dsp syntax
 */
class UnparseVisitor extends DsVisitor {
    visitRule_pool(ctx) {
        const rules = ctx.rule_();
        if (!rules || rules.length === 0) {
            return '';
        }
        return rules.map(r => this.visit(r)).join('\n');
    }

    visitRule(ctx) {
        const terms = ctx.term();
        if (!terms || terms.length === 0) {
            return '';
        }
        
        const result = terms.map(t => this.visit(t));
        const conclusion = result.pop();
        
        if (result.length === 0) {
            return conclusion;
        }
        
        return result.join(', ') + ' -> ' + conclusion;
    }

    visitSymbol(ctx) {
        return ctx.SYMBOL().getText();
    }

    visitSubscript(ctx) {
        const terms = ctx.term();
        if (!terms || terms.length === 0) {
            return '';
        }
        
        const base = this.visit(terms[0]);
        const indices = terms.slice(1).map(t => this.visit(t)).join(', ');
        return `${base}[${indices}]`;
    }

    visitFunction(ctx) {
        const terms = ctx.term();
        if (!terms || terms.length === 0) {
            return '';
        }
        
        const func = this.visit(terms[0]);
        const args = terms.slice(1).map(t => this.visit(t)).join(', ');
        return `${func}(${args})`;
    }

    visitUnary(ctx) {
        const op = ctx.SYMBOL().getText();
        const operand = this.visit(ctx.term(0));
        return `${op} ${operand}`;
    }

    visitBinary(ctx) {
        const op = ctx.SYMBOL().getText();
        const left = this.visit(ctx.term(0));
        const right = this.visit(ctx.term(1));
        return `(${left} ${op} ${right})`;
    }
}

/**
 * Convert from lisp-like Ds syntax to traditional Dsp syntax
 * @param {string} input - Input text in Ds syntax
 * @returns {string} Output text in Dsp syntax
 */
export function unparse(input) {
    const chars = new antlr4.CharStream(input);
    const lexer = new DsLexer(chars);
    const tokens = new antlr4.CommonTokenStream(lexer);
    const parser = new DsParser(tokens);
    const tree = parser.rule_pool();
    const visitor = new UnparseVisitor();
    return visitor.visit(tree);
}
